src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -514,13 +514,26 @@ public override void VisitObjectSyntax(ObjectSyntax syntax)
                     return ErrorType.Create(errors);
                 }
 
+                // Discriminated objects should have been resolved by the declared type manager.
+                var declaredType = typeManager.GetDeclaredType(syntax);
+
                 // type results are cached
                 var namedProperties = syntax.Properties
                     .GroupByExcludingNull(p => p.TryGetKeyText(), LanguageConstants.IdentifierComparer)
-                    .Select(group => new TypeProperty(group.Key, UnionType.Create(group.Select(p => typeManager.GetTypeInfo(p)))));
+                    .Select(group => 
+                    {
+                        var resolvedType = UnionType.Create(group.Select(p => typeManager.GetTypeInfo(p)));
+
+                        if (declaredType is ObjectType objectType && objectType.Properties.TryGetValue(group.Key, out var property))
+                        {
+                            return new TypeProperty(property.Name, resolvedType, property.Flags, property.Description);
+                        }
+
+                        return new TypeProperty(group.Key, resolvedType);
```

Comment:
>TypeProperty [](start = 35, length = 12)

Can you add comments explaining each branch? I'm guessing L532 covers the cases where we have a property that doesn't exist in the types?